[🐛 Bug]: [java] `ExpectedConditions.or` exits early on exceptions

### Description

There appears to be a regression in the java `ExpectedConditions.or` introduced in 81a8322d86ed121e053a6307a42078bdc9129340. Previously while iterating through the conditions, it would catch any `RuntimeException` and save that exception for after checking all conditions. However, that commit changed it so that it only catches `StaleElementReferenceException`. This means if a condition throws something else like `NoSuchElementException`, the iteration will end prematurely and the additional conditions will not be checked. Was this an intentional change?

### Reproducible Code

```shell
@FindBy(id = "fake-id")
private WebElement fakeId;

@Test
public void reproduce() {
    WebDriver driver = webDriver();
    driver.get("https://google.com");
    PageFactory.initElements(driver, this);
    new FluentWait<>(driver)
            .withTimeout(Duration.ofSeconds(10))
            .ignoring(NoSuchElementException.class)
            .until(ExpectedConditions.or(
                    ExpectedConditions.visibilityOf(fakeId),
                    ExpectedConditions.urlContains("google")
            ));
}
```

### Debugging Logs

```logs
org.openqa.selenium.TimeoutException: Expected condition failed: waiting for at least one condition to be valid:
1. visibility of Proxy element for: DefaultElementLocator 'By.id: fake-id'
2. url to contain "google". Current url: ""

(tried for 10 seconds with 500 milliseconds interval)

	at org.openqa.selenium.support.ui.FluentWait.timeoutException(FluentWait.java:276)
	at org.openqa.selenium.support.ui.FluentWait.until(FluentWait.java:231)
	at com.example.ExampleTest.reproduce(ExampleTest.java:86)
	...
Caused by: org.openqa.selenium.NoSuchElementException: no such element: Unable to locate element: {"method":"css selector","selector":"#fake\-id"}
  (Session info: chrome=144.0.7559.133)
For documentation on this error, please visit: https://www.selenium.dev/documentation/webdriver/troubleshooting/errors#nosuchelementexception
Build info: version: '4.40.0', revision: 'b3333f1'
System info: os.name: 'Mac OS X', os.arch: 'x86_64', os.version: '15.7.3', java.version: '17.0.10'
Driver info: org.openqa.selenium.chrome.ChromeDriver
Command: [08f9eb1f2cdaa869b17bddac51ed3b82, findElement {value=fake-id, using=id}]
Capabilities {acceptInsecureCerts: false, browserName: chrome, browserVersion: 144.0.7559.133, chrome: {chromedriverVersion: 144.0.7559.133 (7da823b1d59..., userDataDir: /var/folders/rx/6jz8sptn62q...}, fedcm:accounts: true, goog:chromeOptions: {debuggerAddress: localhost:53562}, goog:processID: 69967, networkConnectionEnabled: false, pageLoadStrategy: normal, platformName: mac, proxy: Proxy(), se:cdp: ws://localhost:53562/devtoo..., se:cdpVersion: 144.0.7559.133, setWindowRect: true, strictFileInteractability: false, timeouts: {implicit: 0, pageLoad: 300000, script: 30000}, unhandledPromptBehavior: dismiss and notify, webauthn:extension:credBlob: true, webauthn:extension:largeBlob: true, webauthn:extension:minPinLength: true, webauthn:extension:prf: true, webauthn:virtualAuthenticators: true}
```

---
ℹ️ **Last known working version:** `4.32.0`

--- a/java/src/org/openqa/selenium/support/ui/ExpectedConditions.java
+++ b/java/src/org/openqa/selenium/support/ui/ExpectedConditions.java
@@ -399,7 +399,7 @@ public class ExpectedConditions {
 
     return new ExpectedCondition<>() {
       private @Nullable String elementText;
-      private @Nullable StaleElementReferenceException error;
+      private @Nullable WebDriverException error;
 
       @Override
       public Boolean apply(WebDriver driver) {
@@ -408,7 +408,7 @@ public class ExpectedConditions {
         try {
           elementText = element.getText();
           return elementText.contains(text);
-        } catch (StaleElementReferenceException e) {
+        } catch (StaleElementReferenceException | NoSuchElementException e) {
           error = e;
           return false;
         }
@@ -481,7 +481,7 @@ public class ExpectedConditions {
 
     return new ExpectedCondition<>() {
       private @Nullable String actualValue;
-      private @Nullable StaleElementReferenceException error;
+      private @Nullable WebDriverException error;
 
       @Override
       public Boolean apply(WebDriver driver) {
@@ -490,7 +490,7 @@ public class ExpectedConditions {
         try {
           actualValue = element.getAttribute("value");
           return actualValue != null && actualValue.contains(expectedValue);
-        } catch (StaleElementReferenceException e) {
+        } catch (StaleElementReferenceException | NoSuchElementException e) {
           error = e;
           return false;
         }
@@ -825,7 +825,7 @@ public class ExpectedConditions {
           // Calling any method forces a staleness check
           element.isEnabled();
           return false;
-        } catch (StaleElementReferenceException expected) {
+        } catch (StaleElementReferenceException | NoSuchElementException expected) {
           return true;
         }
       }
@@ -855,7 +855,7 @@ public class ExpectedConditions {
       public @Nullable T apply(WebDriver driver) {
         try {
           return condition.apply(driver);
-        } catch (StaleElementReferenceException e) {
+        } catch (StaleElementReferenceException | NoSuchElementException e) {
           return null;
         }
       }
@@ -1677,15 +1677,21 @@ public class ExpectedConditions {
    */
   public static ExpectedCondition<Boolean> or(final ExpectedCondition<?>... conditions) {
     return new ExpectedCondition<>() {
+      @Nullable final Object[] results = new Object[conditions.length];
+
       @Override
       public Boolean apply(WebDriver driver) {
-        for (ExpectedCondition<?> condition : conditions) {
+        for (int i = 0; i != conditions.length; ++i) {
+          ExpectedCondition<?> condition = conditions[i];
+          results[i] = null;
           try {
             Object result = condition.apply(driver);
             if (Boolean.TRUE.equals(result) || result != null && !(result instanceof Boolean)) {
               return true;
             }
-          } catch (StaleElementReferenceException ignore) {
+            results[i] = result;
+          } catch (RuntimeException e) {
+            results[i] = e;
           }
         }
         return false;
@@ -1696,9 +1702,22 @@ public class ExpectedConditions {
         StringBuilder message =
             new StringBuilder("at least one condition to be valid:").append(lineSeparator());
         for (int i = 0; i < conditions.length; i++) {
-          message.append(i + 1).append(". ").append(conditions[i]).append(lineSeparator());
+          message
+              .append(i + 1)
+              .append(". ")
+              .append(conditions[i])
+              .append(resultAsString(results[i]))
+              .append(lineSeparator());
         }
         return message.toString();
+      }
+
+      private Object resultAsString(@Nullable Object result) {
+        return Boolean.FALSE.equals(result)
+            ? ""
+            : result instanceof WebDriverException
+                ? " (caused by: " + shortDescription((WebDriverException) result) + ")"
+                : " (actual: " + result + ")";
       }
     };
   }

--- a/java/test/org/openqa/selenium/support/ui/ExpectedConditionsTest.java
+++ b/java/test/org/openqa/selenium/support/ui/ExpectedConditionsTest.java
@@ -966,8 +966,7 @@ final class ExpectedConditionsTest {
           .thenReturn("16pt")
           .thenReturn("17pt")
           .thenReturn("18pt");
-      when(mockElement.getText())
-          .thenThrow(new StaleElementReferenceException("Element disappeared"));
+      when(mockElement.getText()).thenThrow(new NoSuchElementException("Element disappeared"));
 
       assertThat(
               wait.until(
@@ -981,7 +980,7 @@ final class ExpectedConditionsTest {
     void whenAllThrow() {
       String attributeName = "test";
       when(mockElement.getAttribute(attributeName))
-          .thenThrow(new StaleElementReferenceException("Disappeared 1"));
+          .thenThrow(new NoSuchElementException("Disappeared 1"));
       when(mockElement.getCssValue(attributeName))
           .thenThrow(new StaleElementReferenceException("Disappeared 2"));
       when(mockElement.getText()).thenThrow(new StaleElementReferenceException("Disappeared 3"));
@@ -999,7 +998,8 @@ final class ExpectedConditionsTest {
                   + "1. element to have text \"test\", but..."
                   + " org.openqa.selenium.StaleElementReferenceException: Disappeared 3."
                   + lineSeparator()
-                  + "2. attribute or CSS value \"test\"=\"test\". Current value: \"null\".")
+                  + "2. attribute or CSS value \"test\"=\"test\". Current value: \"null\". (caused"
+                  + " by: org.openqa.selenium.NoSuchElementException: Disappeared 1)")
           .hasMessageContaining("tried for 1.1 seconds with 250 milliseconds interval");
     }
   }

